### PR TITLE
Feature: Verify login on signup with magic link.

### PR DESCRIPTION
**Pull request by agustif**
Needed for calcom/website#93

## What does this PR do?
This PR is part of a bigger effort, check the PR in website repo for more info on how to test both.

## Changes
Update verificationRequest -> verificationToken in:
- [packages/prisma/schema.prisma](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-56fa4d384ba6fb94b130e9eb42c5cad00734b025ebd6c197f76873469725ae87)
- [apps/web/pages/api/teams/[team]/invite.ts](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-f891cc16d5b2e73b7e04c72f3ce56eafe32fbf3943c68967666f7293bdfb34a3)
- [apps/web/pages/auth/signup.tsx](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-51d962487807b28a7b251dac5e5e3ce33f4e2e1d0bd2ffe6e2e0b5a42ef24c93)
- [apps/web/playwright/auth/auth-index.test.ts](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-c5c34ad68119913ba6093b5f91cc70986a0c9d82a38a70573826d6709c12f801)

Adds two utils (default-cookies and env) at packages lib:
- [packages/lib/default-cookies.ts](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-a6b2fe08300f1b2178eb4c81bbcc11b259326189526fedd23d6af1d3fd3dd2ae)
- [packages/lib/env.ts](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-7cc2b5b74e58bea1497d327ebaabd2f5ddfafec5d18cf19f6453acd1c63eb260)

Adds a new SQL/ prisma migration
- [...igrations/20220306010113_renames_verification_request_to_verification_token/migration.sql](https://github.com/calcom/cal.com/compare/feat/verify-login?expand=1#diff-5fb64ec326bdcba32526e5769863bfcfee8cb3238dd7d8b04b43b92b073d6d70)

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?
Please refer to the PR in website for more info on testing.

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by sean-brydon — APPROVED** [review]
LGTM :)

**Review by PeerRich — CHANGES_REQUESTED** [review]
needs to be fixed: https://github.com/calcom/website/pull/82#issuecomment-1066094525

File: apps/web/lib/emails/templates/attendee-scheduled-email.ts [hunk 1]
```diff
@@ -6,11 +6,12 @@ import utc from "dayjs/plugin/utc";
 import { createEvent, DateArray } from "ics";
 import nodemailer from "nodemailer";
 
+import { serverConfig } from "@calcom/lib/serverConfig";
+
```
Comment (agustif) [inline apps/web/lib/emails/templates/attendee-scheduled-email.ts:10]: I moved this from web/lib to packages/lib so website app can call it too. all the email templates are updated

File: packages/lib/default-cookies.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * For more on prefixes see https://googlechrome.github.io/samples/cookie-prefixes/
+ *
+ */
+
+export function defaultCookies(useSecureCookies: boolean): CookiesOptions {
+  const cookiePrefix = useSecureCookies ? "__Secure-" : "";
+  // To enable cookies on widgets,
+  // https://stackoverflow.com/questions/45094712/iframe-not-reading-cookies-in-chrome
+  // But we need to set it as `lax` in development
+  const sameSite = isENVDev ? "lax" : "none";
```
Comment (agustif) [inline packages/lib/default-cookies.ts:22]: This was working, but we need to make sure we're using securized cookies on prod, I use this isENVDev to detect the envirorment, also defaultCookies accepts a boolean to useSecureCookies or not. passing the current URL to it with startsWith(https://...) does the trick

File: packages/lib/serverConfig.ts [hunk 3]
```diff
@@ -16,6 +18,9 @@ function detectTransport(): SendmailTransport.Options | SMTPConnection.Options |
         pass: process.env.EMAIL_SERVER_PASSWORD,
       },
       secure: port === 465,
+      tls: {
+        rejectUnauthorized: isENVDev ? false : true,
+      },
```
Comment (agustif) [inline packages/lib/serverConfig.ts:23]: This was necessary for me to be able to run maildev and recieve mails on localhost, it only disables rejecting unathorized in dev, not prod.

File: packages/prisma/migrations/20220306010113_renames_verification_request_to_verification_token/migration.sql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+ALTER TABLE IF EXISTS "VerificationRequest" RENAME TO "VerificationToken";
+
+-- RenameIndex
+ALTER INDEX IF EXISTS "VerificationRequest_pkey" RENAME TO "VerificationToken_pkey";
+
+-- RenameIndex
+ALTER INDEX  IF EXISTS "VerificationRequest_token_key" RENAME TO "VerificationToken_token_key";
+
+-- RenameIndex
+ALTER INDEX IF EXISTS "VerificationRequest_identifier_token_key" RENAME TO "VerificationToken_identifier_token_key";
```
[the remark this answers is not in the document]
Comment (agustif) [reply packages/prisma/migrations/20220306010113_renames_verification_request_to_verification_token/migration.sql:10]: But the original ones prisma dev generated did make less sense dropping and recreating tables an lots of indices

**Review by agustif — COMMENTED** [review]
I think this is ready, but please do check the cookies security part thoroughly!

**Review by zomars — COMMENTED** [review]
This is a great example of a self-reviewed PR. Thank you for this. Will deploy this branch to our staging domain so we can actually test it on a prod-like environment.

**Review by PeerRich — APPROVED** [review]

**Comment by PeerRich** [comment]
after doing some testing i found quite a few problems

- wrong email template (please use our existing one)
- issues with premium sign up (ask for verified email *after* the stripe checkout)
- (NIT: lets use the toast design we have in the app)
- one time i signed up and it didnt redirect to cal.dev/new and just showed the toast without doing anything
- cal.dev/new still redirects to the auth screen without signing in

**Review by PeerRich — DISMISSED** [review]
after doing some testing i found quite a few problems

- wrong email template (please use our existing one)
- issues with premium sign up (ask for verified email after the stripe checkout)
- (NIT: lets use the toast design we have in the app)
- one time i signed up and it didnt redirect to cal.dev/new and just showed the toast without doing anything
- cal.dev/new still redirects to the auth screen without signing in

File: packages/prisma/migrations/20220306010113_renames_verification_request_to_verification_token/migration.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (alannnc) [reply packages/prisma/migrations/20220306010113_renames_verification_request_to_verification_token/migration.sql:10]: Probably was due to a db sync problem, but after that it should give you exactly what you needed.

**Review by alannnc — APPROVED** [review]
Code looks good, what do I need to test it locally? @agustif

**Comment by agustif** [comment]
> Code looks good, what do I need to test it locally? @agustif

Hey @alannnc you need to checkout this other PR on website and run both at the same time. in the other one there's more context on how to run locally!

File: apps/web/components/Shell.tsx [hunk 5]
```diff
@@ -69,7 +69,7 @@ function useRedirectToLoginIfUnauthenticated(isPublic = false) {
       router.replace({
         pathname: "/auth/login",
         query: {
-          callbackUrl: `${WEBAPP_URL}/${location.pathname}${location.search}`,
+          callbackUrl: `${WEBAPP_URL}${location.pathname}${location.search}`,
```
Comment (zomars) [inline apps/web/components/Shell.tsx:72]: Prevents double slash

**Comment by hariombalhara** [comment]
Email Provider doesn't seem to be working for me @agustif . sendVerificationEmail function is not called.

**Comment by hariombalhara** [comment]
I was able to fix the Email Provider issue. But /new endpoint is not working.

**Comment by zomars** [comment]
> * issues with premium sign up (ask for verified email after the stripe checkout)

This will be handled in the new "In-app" email verification flow @PeerRich @agustif

File: apps/web/pages/api/auth/[...nextauth].tsx [hunk 6]
```diff
@@ -141,15 +154,50 @@ if (isSAMLLoginEnabled) {
   });
 }
 
+if (true) {
```
Comment (alannnc) [inline apps/web/pages/api/auth/[...nextauth].tsx:157]: This has no point, as its always being executed

Comment (agustif) [reply apps/web/pages/api/auth/[...nextauth].tsx:157]: I believe it's in order to remind us that we might want to disable it in some instances in the future or allow to config it? not sure!

Comment (zomars) [reply apps/web/pages/api/auth/[...nextauth].tsx:157]: I just liked to encapsulate the provider logic, also as @agustif we can disable easily maybe with an `env` variable.

**Pull request merged**